### ISSUE TITLE
Empty author name issue resulting from empty user first and last names

### ISSUE THREAD

**Issue by hyi** [hydroshare/hydroshare#3991]
Need to make first and last names mandatory fields when a user signs up to fix this issue.

**Comment by hyi** [comment]
@danames @dtarb @martinseul I looked into the code and found the reason first name and last name on user signup page are optional is that we are using the default Django User model that includes username, first_name, last_name and email and the default Django User model has username and email fields as required but first_name and last_name as optional. So it is not that straightforward to make first_name and last_name required since we would have to move first_name and last_name to UserProfile model rather than using the default Django User model to manage these fields. After giving this a bit of more thought, I think forcing users to input first_name and last_name might be too restrictive anyway. I am proposing instead of always using first_name and last_name to display Authors on the resource landing page, we should use username to represent the author if first_name and last_name are empty. What do you think?

**Comment by dtarb** [comment]
@hyi I think that is a cop out.  I think we need to find a way to have required fields be required fields, and our naming model has firstname and lastname as required fields.   See https://stackoverflow.com/questions/8466851/making-first-name-last-name-a-required-attribute-rather-than-an-optional-one-in for some options.

We have in the past discussed our people name model quite a lot (#45, #855) , and sometimes I have suggested revisiting it to remove inconsistencies such as our resource author name model just having one name field while our user model has firstname, only one middle initial and lastname.  There may be merit in revisiting these questions, but I think that to address the current concern that we should just enforce that there be a first and last name.

**Comment by danames** [comment]
I think I may agree with @hyi  on this one. First and Last names are sort of arbitrary given all the different naming conventions people use in different cultures. I assume that's why django has not made them mandatory. Also for the purposes of an organization joining HydroShare it would be nice to be able to keep the names blank. Though I don't like using "username" for the author. But using "Organization" makes a lot of sense. "Organization" is required in our django data model... 
![image](https://user-images.githubusercontent.com/3472628/95229128-6e8ed580-07bd-11eb-9d68-5409aa3b1099.png)

**Comment by hyi** [comment]
@danames Agreed Organization is much better than username to represent users with empty first and last names for Authors. However, Organization was optional initially and changed to required later with a popup reminder for existing users to fill in organizations if not filled in already, so there is a possibility that a small portion of existing users may still have organizations empty, in which case we could use username as the last resort if both names and organizations are empty. I'll wait for the follow-up weigh-in from @dtarb before starting implementation.

**Comment by horsburgh** [comment]
@hyi and @danames - we already allow Organizations to be Authors of resources. The "correct" way to do this is by adding a new author to the resource as an organization on the "Add Author" dialog. This will ensure that the authorship and the citation are created and displayed correctly on the resource landing page. If you want to have an "organizational" account for HydroShare, I would recommend NOT using it as an author of a resource because the name will not be displayed correctly in the authorship and citation on the landing page. Ownership of resources by an organizational account should be fine. Also - you should consider using a HydroShare Group rather than an organizational user account as our user accounts are designed around individual people.

Given the amount of thought that went into our user/people model - I'm with @dtarb on requiring names for users.

![image](https://user-images.githubusercontent.com/3485902/95284308-20102400-081b-11eb-908d-e94f1a8e2bcd.png)

**Comment by danames** [comment]
I think there will be a lot of cases where resources should be "owned" by
an organization. Can a "group" own a resource? If that's the case, then I
think we can delete our organiational "individual" account and create a
group one.

However, I see "Geotrust CDM" "CZO Shale Hills" "CZO National" and
other organizations appear to have "individual" accounts. So our
instructions and/or messaging is not working?

Looks like they used "CZO" for first name and "National" for last name
which seems wrong...
**********************************************************************
Daniel P. Ames, Ph.D., P.E.
Professor, Civil & Environmental Engineering,Brigham Young University,
Provo, Utah, USA
President, International Environmental Modelling & Software Society
Editor-in-Chief, Environmental Modelling & Software
http://ceen.et.byu.edu/faculty/person/Daniel-P.-Ames
<http://ceen.et.byu.edu/content/dan-ames>
http://www.researchgate.net/profile/Daniel_Ames
http://scholar.google.com/citations?user=S0GUCeUAAAAJ

**********************************************************************

**Comment by dtarb** [comment]
I’ve worked this through and the following is a lengthy post for how I think this should be resolved.  I’ve tried to be thorough, to avoid further band aids.

HydroShare has users and authors.  These are separate concepts.  Users may own a resource.  When a user creates a resource they become the owner and their profile information is copied (a convenience function) into the author records for the resource as they are often the author.  They may then edit the author records if they want to add or change authors.  (Technically authors are creators in Dublin Core terminology).

There is an established need for users to be individuals as well as organizations.  This is manifested by a CUAHSI organizational account as well as many CZO organizational accounts.  There is not the concept of “groups” owning anything (and I suggest we keep it that way).

The user model includes the following fields
-	Firstname
-	Lastname
-	Organization
-	email
-	(others not germane to this discussion not listed) 

The author model includes options for Person and Organization.  For person it includes the following fields
-	Name*
-	Organization
-	Email
-	(others not germane here not listed)

For Organization it includes
-	Organization*
-	Email
-	(others not germane here not listed)

There are the following inconsistencies
-	The sign up page for users does not require Firstname and Lastname to be entered, however user profile editing does require these fields. 
-	The user model does not naturally support organizational accounts
-	The convenience function mapping from the user model to author model creates a blank author name entry for user accounts without firstname and lastname
-	The resource landing page does not enable opening and thus editing of authors where name is present but blank, as created by a user who has blank firstname and lastname.  Thus it is not possible for a user to correct this author problem on an existing resource.

Suggested solution
1.	Drop the requirement for First and Last names to be given.  This better supports organizations and also protects privacy (we had one user who did not want their name to be on a public website).  User accounts are uniquely identified by their email address with the Organization field serving as an attribute, and Firstname and Lastname as additional attributes if provided.  This follows the rationale that an account is validated by the user responding to email received at the email address given when creating an account.  A user account in HydroShare can be for whatever entity creates it, person or Organization.  
2.	Correct the convenience function that writes resource authors to handle the case where user firstname and lastname are blank.  In these cases create an Organization author record, with the Author name the Organization name.  A user will have the ability to edit this if needed.  
3.	Add a new user type called “Organization” to our current list of user types.  The user type field is already required and set to be chosen from a fixed list.  [This is not essential for fixing problems, but I think is needed because organization accounts do not neatly fit into existing categories.]

Some further justification.
This is presently a problem for the CUAHSI organizational account where the profile is 
![image](https://user-images.githubusercontent.com/3456833/95661504-1faeac00-0aed-11eb-96b4-2bde39699b39.png)
It is not possible at present for the CUAHSI user to fix this without entering both a firstname and lastname.  It is also not possible for an administrator to fix this.

There are CZO users with profiles such as
![image](https://user-images.githubusercontent.com/3456833/95661512-2c330480-0aed-11eb-9c0e-83e6c6ece1f3.png)
This has
Firstname: CZO
Lastname: IML
Organization: CZO

If this user creates the resource and does not actively edit the Author record, the citation will be
IML, C, (year). “title” …

If it was purely and Organization user, with better name “Intensively Managed Landscapes Critical Zone Observatory” this would go on the Author Organization field and be cited
Intensively Managed Landscapes Critical Zone Observatory, (year). “title”


Following are details.  There are multiple issues.  @hyi, if you want to split this into separate issues for implementation, feel free.

### First Issue
Relax requirements on Firstname and Lastname.
Sign up page
![image](https://user-images.githubusercontent.com/3456833/95661564-8df36e80-0aed-11eb-8c6e-ebc762aa38c5.png)
Note that this change suggests the need for “Organization” to be added as an allowed user type. 

User profile editing page
![image](https://user-images.githubusercontent.com/3456833/95661573-99469a00-0aed-11eb-89c4-32052aa98de1.png)
### Second Issue
Resources where the Author is missing.  This is caused by the first issue of a resource having an author that is empty.  I was able to create an account without names to test this out.  This shows up as below in view mode
![image](https://user-images.githubusercontent.com/3456833/95661578-a794b600-0aed-11eb-8e56-f5d0d81ef411.png)
And in edit mode
![image](https://user-images.githubusercontent.com/3456833/95661585-b3807800-0aed-11eb-9e41-8d2ad345e576.png)
In the bag metadata we have the following where the Dublin Core creator is blank (and as a reminder the Dublin Core creator term we label as author on our interface)
![image](https://user-images.githubusercontent.com/3456833/95661587-bed3a380-0aed-11eb-991f-523b88a4ca32.png)

When a resource is created, the author information is copied from the profile of the user creating the resource.  This is a convenience function, because often the user creating the resource is the author.  But this does not have to be the case.  Authors on resources can then be added to, removed, and edited.  The “bug” here is that there is no text to click on to edit the blank author.  

I suggest that to fix this, the convenience function that creates authors from owners be changed to create an Organization author when the owner user does not have firstname or lastname.  Change the Django and bag metadata so that the creator entry is

![image](https://user-images.githubusercontent.com/3456833/95661594-ceeb8300-0aed-11eb-872b-1b6206bcdfa4.png)

Then the Author created becomes an organization Author.  The basis for thinking this, is that if I add an organization author to a resource I get the following (without a name term)

![image](https://user-images.githubusercontent.com/3456833/95661599-d743be00-0aed-11eb-9409-9364960ba344.png)

In the implementation of this, there will need to be a migration or management command that finds all resources where hsterms:name is blank and deletes the name entry from the creator record, so that these creator records become Organizations.

An issue to address here is what to do when there are blank first and last names and multiple organizations.  Right now the creator record gets set up like
![image](https://user-images.githubusercontent.com/3456833/95661613-f17d9c00-0aed-11eb-9a37-49200920d607.png)
I suggest that we do nothing, except remove the blank hsterms:name entry.  This treats the list of organizations as one organization with a long name.  I debated creating separate Organization author entries for each, but decided against this because they are all associated with a single organization user.  It is quite easy for a resource editor to change these author entries as needed.
![image](https://user-images.githubusercontent.com/3456833/95661638-183bd280-0aee-11eb-9b88-fee62fc1aada.png)

### Third Issue
If I add “Other person” author to a resource, the Name field is indicated as required.  However, this is not enforced.  If I enter only an organization (e.g. “uwrl”)
![image](https://user-images.githubusercontent.com/3456833/95661644-22f66780-0aee-11eb-8e11-d19b520fce46.png)
I get uwrl as an author 
![image](https://user-images.githubusercontent.com/3456833/95661646-2db0fc80-0aee-11eb-917b-7f2c0e4d6e32.png)
But if I go to view mode I get “null” instead of UWRL
![image](https://user-images.githubusercontent.com/3456833/95661652-386b9180-0aee-11eb-802f-52d9e1872f07.png)
If I go back to edit mode, I still see null.
If I download the bag and look at the bag metadata I see
![image](https://user-images.githubusercontent.com/3456833/95661656-40c3cc80-0aee-11eb-9e9b-ae481581c3ab.png)
Null was automatically entered, but evidently not displayed until switching to view mode.
I suggest the fix is to not allow save changes on the Add Author form if Name has not been filled in, as Name is indicated as required.  If a user wants to create an Organization author entry, they should use the Organization option, not Other Person option.
### Fourth issue
I suggest allowing an administrative user to edit the profile of any user.  This will be helpful for editors to help users fix up profiles that are incorrect.  This will require adding the edit button for administrator users when viewing any user profile so they can access it in edit mode.
![image](https://user-images.githubusercontent.com/3456833/95661670-533e0600-0aee-11eb-9dd1-8e5eccaba095.png)
### Fifth issue
Metadata for user with space in last name.  If I have a space in a users last name, such as
![image](https://user-images.githubusercontent.com/3456833/95661681-60f38b80-0aee-11eb-9dad-f352cbddf736.png)
I get the following under how to cite
![image](https://user-images.githubusercontent.com/3456833/95661690-66e96c80-0aee-11eb-90d8-a0ee2307a398.png)
This is correct.  If I Edit Author, I see the comma following Tarboton, indicating that “Drennan Tarboton” is the lastname and Debbie first name for citation purposes.
![image](https://user-images.githubusercontent.com/3456833/95661695-6fda3e00-0aee-11eb-93fb-73cd71971f03.png)
However, if I look at the bag metadata I see
![image](https://user-images.githubusercontent.com/3456833/95661700-79fc3c80-0aee-11eb-9107-6ba8f52d11d2.png)
There is no indication where the firstname ends and lastname begins in hsterms:name.  Thus bag importing has no way to get this right. 

The text for this comment is also in https://docs.google.com/document/d/1cHHC55jN4srMMCeWXtIWhjTMXe-EA50d/edit#

**Comment by horsburgh** [comment]
@dtarb - regarding the first issue - could we move the "User Type" drop down to the top of the form and then make the form respond to what they select? The issue is - if someone selects a user type of "Organization" but then enters a first and last name, what should the system do with that? I suggest that if the person selects a user type of "Organization" then the name fields should not appear on the form for entry and they only have the option to enter the "Organization". You might say that an organization should have a contact person. If that's the case, when a user of type "Organization" authors content in HydroShare, is the organization's name or the contact person's name listed as the Author? If we keep the first and last name fields for users of type organization, then I suggest the organization name is listed as the author when that user creates content.

Also, on the existing form, the "Organization(s)" text entry box implies that there can be multiple organizations entered. For a user of type "Organization" this would not seem to be the desired behavior. I suggest modifying that on the form so that only one organization name is entered for a user type of "Organization". We could leave this as is, but it will likely result in weird entries.

Finally - What is the purpose of a user account with no first and last name? I understand that some may not want their names to appear on a public website. That is fine. They can choose not to sign up for an account, and they can choose not to share content in HydroShare. I'm just wondering what a user should be able to do with a user account that has no name? Doesn't that mean that they are essentially the same as an anonymous, unauthenticated user?

**Comment by dtarb** [comment]
@horsburgh I am happy if you want to flesh out and specify an alternative approach.  I was trying to develop an approach that was simple and had the least disruption to what we have.  The approach you are suggesting is certainly viable, but seems to add additional logic to the sign up and profile editing forms.

With respect to no first and last name, there was a case of a user who signed up for an account with an obviously fake name.  I chided them over not using their real name and they came back saying they had been in the military and did not want their real name on any public website due to security concerns.  You may say that we do not check fake names, so using fake names is OK.  You said above they have the choice as to whether to get an account.  In their case they were required to get an account to do homework of a university class that was using HydroShare (a practice we should value).  So while certainly a user can put something in for firstname and lastname, allowing them to be left blank meets that need directly.

**Comment by hyi** [comment]
@dtarb @horsburgh @danames Creating a new user type of "Organization" is a bit confusing to me since (1) Each user can be associated with multiple organizations which is also a required field on user signup page; (2) All other user types are associated with an individual user, so a "Organization" user type seems to be an "outlier" to me among other supported user types; (3) Organization seems to me to fit more as a group rather than as a user. Can czo users be created as one czo group with multiple "normal" users in the czo group? Somehow treating/fitting an "Organization" into our user model feels a bit confusing to me.

**Comment by hyi** [comment]
@dtarb One question to confirm with you regarding your fourth issue: allow admin to edit profile of any users. While working on this, I found our code has assumed the requesting user (which would be admin in this case) is the user whose profile to be updated and this logic is used almost everywhere in our code involving profile update. To allow admin to edit profile of any users, I have to update the logic everywhere, so I am questioning a bit on whether the benefit of allowing this is worth the effort. Also, when an admin logs in, there is no link that would take the logged in admin to the profile page of a user the admin wants to update, and the admin has to fine the user's id and manually types in the URL of the profile page of the user in order to get to the user profile page to update other user's profile. So I am also not sure how much the admin will really use this functionality to update another user's profile. I'm leaning toward not implementing this given the work being involved with little value-add, but want to get your thought.

**Comment by dtarb** [comment]
I am OK putting issue 4 on an enhancement list to consider for later.

There are many ways to get to the profile page of a user.  E.g. if you click on their name in Discover for a resource they are author of or their name in a group, or their name on a resource landing page where they are an author, or owner.  Note that the admin user can use HydroShare just like a regular user.  Once at the profile page I was just suggesting there be an edit button that could then take them to editing.  This is exactly the same pattern that an admin user can use to edit a resource.
![image](https://user-images.githubusercontent.com/3456833/97656025-09c94280-1a2c-11eb-8182-0c4cf9474e83.png)

vs
![image](https://user-images.githubusercontent.com/3456833/97656081-341b0000-1a2c-11eb-9a6d-79b3ab4aa818.png)

If you have implemented other items and are deferring on issue 4, please put it in a separate issue before closing this issue.

**Comment by hyi** [comment]
@dtarb OK. Thanks for the info. I'll take a stab at this first while I am at it then.

**Comment by dtarb** [comment]
I tested the deployment of the branch to fix this issue on dev-hs-3.cuahsi.org.

I think that issues 1, 2, 4 and 5 are essentially fixed.  Good work!  On issue 1 there are some additional labeling changes I would like.  See below.

I found the following problem with issue 3.  If I try to add an Organization Author to a resource, I get an error “Name is required” even though there is not a name field to add.
![image](https://user-images.githubusercontent.com/3456833/98129527-743c1180-1e76-11eb-9e95-5544438e7c56.png)

Also, when this issue was first written we forgot to note that Contributors should be treated identically as Authors to the extent possible.  Thus we should not have the following in metadata when there is an organization contributor.
![image](https://user-images.githubusercontent.com/3456833/98129579-83bb5a80-1e76-11eb-8d6f-b8f238e4299c.png)
The contributors area on a resource landing page should say 
“People or Organizations that contributed …”
![image](https://user-images.githubusercontent.com/3456833/98129627-946bd080-1e76-11eb-9bc8-81b3b69e1b61.png)

Also – to avoid the awkwardness of name and organization for organization contributors I suggest a layout that just has the column header “Contributor” as indicated that combines name and organization entries, separated by a comma.  Omit the comma if one of the fields is empty.  Also add an active hyperlink if it exists.  There are two possible links.  I suggest that if there is a homepage link that it be given preference, and that the hydroshare profile link only be used if there is no homepage link.

The add contributor dialog, just like the add author dialog should include organization contributors.

@horsburgh may want to comment on this.

Some small refinements
On the sign up page I think it should say “Names may be omitted if you do not wish to provide them or are creating an Organization user account.”  I.e. expand the text below.
![image](https://user-images.githubusercontent.com/3456833/98129739-b5ccbc80-1e76-11eb-9b24-75eea1e268b6.png)

Also add on the profile editing page below First name, middle name, Last name
“If this is an account for an Organization User, First name, Middle name and Last name may be left blank.  Resources created by this user will then be created with the Organization for the resource author/creator.” 
![image](https://user-images.githubusercontent.com/3456833/98129797-c7ae5f80-1e76-11eb-86ee-ade2f96c526d.png)

Also add “Organization” to the list below for users to indicate that the account is representing an organization.  @Castronova @martinseul  may want to comment on this.
![image](https://user-images.githubusercontent.com/3456833/98129877-df85e380-1e76-11eb-9d6f-364353a2be13.png)

**Comment by hyi** [comment]
@dtarb I should be able to take care of other issues easily except for the Organization Author. While trying to fix these issues, I found Organization Author is implemented as a Creator without names. I thought this should suffice without introducing an "Organization user" on the user sign up page. We may discuss the potential value-add of having an organization user, since I don't yet see the value-add especially after I have fixed these related empty author name issues.

**Comment by horsburgh** [comment]
I would change the text added below First name, Middle name, and Last Name to the following:

> If this user account is for an Organization, First name, Middle name, and Last name may be left blank. Any resources created by an Organization user will list the Organization as the resource author/creator.

Otherwise, I think these suggested changes seem fine.

@hyi - we have at least several "Organization users" in HydroShare now. People are setting them up to have a user account that can own and manage a bunch of resources across an organization - even if that Organization user is not the author/creator of those resources. I think it's just recognizing that people are already using the system in this way.

**Comment by hyi** [comment]
@horsburgh Yeah, I think I get your point. With the fixes I made, when a user signs up without providing first and last names, the user has to fill up the organization required field, and this organization is used throughout HydroShare now to represent the user. From the Creator metadata element perspective, this user without name is managed in the same way as an Organization Author being added to an existing resource. So I thought when a user wants to sign up as an organization user, the user just need to leave the name blank and fill in the organization required field and we don't need to add an "Organization" type to the user type dropdown. We can make the explanation text below the first and last name clearer to cover this case. My point is that we don't need to add an "Organization" type to explicitly mark the user without name as such, and we can treat all users without names as Organization user, which might be sufficient to meet our needs.

**Comment by dtarb** [comment]
@hyi I agree with @horsburgh wording changes.  The suggestion for an "Organization" to be on the list of user types is just an issue of labeling.  It is not intended to impact functionality at all.  If you are creating a user account to, for example be used by RENCI to own and be an author for resources that RENCI shares or publishes, RENCI does not really fit into any of the categories in the user type list.  It is just to provide a category for Organizations to fit in to.  It may be used in reporting (we have this number of types of users), but not for any actual website functionality.

**Comment by hyi** [comment]
@horsburgh @dtarb Adding an Organization type in the dropdown was a bit confusing to me initially, but after thinking this a bit more with the help of your clarifications, it starts to make sense to me. I'll address this as you suggested then. Thanks for the clarifications.

**Comment by dtarb** [comment]
@hyi I reviewed changes again.  Two small things
- I think "Organization" should be near the bottom (just above other) in user type.  Not first.  It is a rare case and we do not want it chosen by mistake.
- Is it possible to avoid the indent wrapping here
![image](https://user-images.githubusercontent.com/3456833/98196114-8a7eb780-1ee0-11eb-9fb3-f16de6662cd7.png)

Nice touch to switch to Organization user type automatically when names are blank in profile editing.

**Comment by hyi** [comment]
@dtarb Have updated the cuahsi test VM with fix to these two small things. Please take a quick look when you get a chance and let me know if you notice any further issues.

**Comment by dtarb** [comment]
@hyi I am happy with the changes.

**Comment by hyi** [comment]
Deployed. Closing.
